src/javascript/request-composer/request.js

```diff
@@ -0,0 +1,318 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import moment from 'moment';
+
+function buildReportRequest(params) {
+  let reportRequest = {
+      viewId: params.viewId,
+      samplingLevel: params.samplingLevel,
+  };
+  if (params.filters) {
+    reportRequest.filtersExpression = params.filters;
+  }
+  if (params.includeEmptyRows) {
+    if (params.includeEmptyRows.toLowerCase() == 'true') {
+      reportRequest.includeEmptyRows = 'true';
+    } else if (params.includeEmptyRows.toLowerCase() == 'false') {
+      reportRequest.includeEmptyRows = 'false';
+    }
+  }
+  if (params.pageSize &&
+      parseInt(params.pageSize)) {
+    reportRequest.pageSize = params.pageSize;
+  }
+  if (params.pageToken) {
+    reportRequest.pageToken = params.pageToken;
+  }
+  return reportRequest;
+}
+
+function applyDateRanges(request, params, settings) {
+  if (params.startDate &&
+      params.endDate &&
+      settings.requestType != 'COHORT') {
+    request.dateRanges = [{
+      'startDate': params.startDate,
+      'endDate': params.endDate
+    }];
+  }
+  return request;
+}
+
+function applyMetrics(request, params, settings) {
+  if (params.metrics &&
+      settings.requestType != 'COHORT') {
+    request.metrics = [];
+
+    let metrics = params.metrics.split(',');
+    for (let i in metrics) {
+      request.metrics.push({'expression': metrics[i]});
+    }
+  } else if (params.cohortMetrics &&
+             settings.requestType == 'COHORT') {
+    request.metrics = [];
+
+    let metrics = params.cohortMetrics.split(',');
+    for (let i in metrics) {
+      request.metrics.push({'expression': metrics[i]});
+    }
+  }
+  return request;
+}
+
+function applyDimensions(request, params, settings) {
+  if (settings.requestType != 'COHORT' &&
+      (params.histogramDimensions || params.dimensions)) {
+    request.dimensions = [];
+    let dimensions = [];
+    if (settings.requestType == 'HISTOGRAM' &&
+        params.histogramDimensions) {
+      dimensions = params.histogramDimensions.split(',');
+    } else if (params.dimensions) {
+      dimensions = params.dimensions.split(',');
+    } else {
+      return request;
+    }
+    for (let i in dimensions) {
+      let dimension = {'name': dimensions[i]};
+      if (settings.requestType &&
+          settings.requestType == 'HISTOGRAM' &&
+          params.buckets) {
+        let histogramBuckets = [];
+        let buckets = params.buckets.split(/[ ,]+/);
```

Comment:
Nit: splitting on `/[ ,]+/` would allow the input to be something like "0,,,10,,,20" or "0 10 20". You could do `/\s*,\s*/` if you wanted to force a single comma.